catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java

```diff
@@ -126,104 +216,196 @@ public boolean onOptionsItemSelected(MenuItem item) {
+			ProjectLoadTask
+					.task(name, this);
+			project = ProjectManager.getInstance().getCurrentProject();
 		}
 
-		Project project = ProjectManager.getInstance().getCurrentProject();
+		project.setDescription(description);
+		project.setDeviceData(this);
+
+		File projectDir = new File(DEFAULT_ROOT_DIRECTORY, project.getName());
 
-		if (Utils.isDefaultProject(project, this)) {
-			nameInputLayout.setError(getString(R.string.error_upload_default_project));
+		final SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(this);
+		String token = sharedPreferences.getString(Constants.TOKEN, Constants.NO_TOKEN);
+		String username = sharedPreferences.getString(Constants.USERNAME, Constants.NO_USERNAME);
+
+		Intent intent = new Intent(this, ProjectUploadService.class);
+
+		String[] sceneNames = project.getSceneNames().toArray(new String[0]);
+
+		final int notificationId = StatusBarNotificationManager.getInstance()
+				.createUploadNotification(this, name);
+
+		uploadProgressDialog = new AlertDialog.Builder(this)
+				.setTitle(getString(R.string.upload_project_dialog_title))
+				.setView(R.layout.dialog_upload_project_progress)
+				.setPositiveButton(R.string.progress_upload_dialog_show_program, null)
+				.setNegativeButton(R.string.done, new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int which) {
+						finish();
+					}
+				})
+				.setCancelable(false)
+				.create();
+
+		uploadProgressDialog.show();
+		uploadProgressDialog.getButton(AlertDialog.BUTTON_POSITIVE).setEnabled(false);
+
+		intent.putExtra("uploadName", name);
+		intent.putExtra("projectDescription", description);
+		intent.putExtra("projectPath", projectDir.getAbsolutePath());
+		intent.putExtra("username", username);
+		intent.putExtra("token", token);
+		intent.putExtra("provider", NO_OAUTH_PROVIDER);
+		intent.putExtra("sceneNames", sceneNames);
+		intent.putExtra("notificationId", notificationId);
+
+		startService(intent);
+		new UploadStatusPollingTask()
+				.execute();
+
+		int numberOfUploadedProjects = sharedPreferences.getInt(NUMBER_OF_UPLOADED_PROJECTS, 0) + 1;
+		sharedPreferences.edit()
+				.putInt(NUMBER_OF_UPLOADED_PROJECTS, numberOfUploadedProjects)
+				.commit();
+
+		if (numberOfUploadedProjects != 2) {
 			return;
 		}
 
-		if (ProjectRenameTask.task(project.getName(), name)) {
-			ProjectLoadTask.task(name, this);
-			project = ProjectManager.getInstance().getCurrentProject();
-		}
+		new AlertDialog.Builder(this)
+				.setTitle(getString(R.string.rating_dialog_title))
+				.setView(R.layout.dialog_rate_pocketcode)
+				.setPositiveButton(R.string.rating_dialog_rate_now, new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int id) {
+						try {
+							startActivity(new Intent(Intent.ACTION_VIEW,
+									Uri.parse("market://details?id=" + getPackageName())).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
```

Comment:
could "market://details" also be defined as constant?

Reply:
yes, it could, but what's the benefit? it's never used anywhere else.